### PR TITLE
yolox related code added

### PR DESCRIPTION
**Pull request by shaydeci**
4th PR as part of Yolox integration in SG- holds Yolox architecture and loss related code parts.

Has some intersection with https://github.com/Deci-AI/super-gradients/pull/207 as the yolo new class structure must be included, so the above needs to be merged first.

### CODE REVIEW

File: tests/unit_tests/yolox_unit_test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                self.assertIsNotNone(output_train)
+
+                # THIS SHOULD RUN THE FORWARD AUGMENT
+                yolo_model.eval()
+                yolo_model.augmented_inference = True
+                output_augment = yolo_model(dummy_input)
+                self.assertIsNotNone(output_augment)
+
+                # MODEL IS ANCHORLESS
+                assert yolo_model._head.anchors.num_anchors == 0
```
Comment (ofrimasad) [inline tests/unit_tests/yolox_unit_test.py:43]: please use `self.assert_equal`

File: tests/unit_tests/yolox_unit_test.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import unittest
+
+import torch
+
+from super_gradients.training.models.detection_models.yolox import YoloX_N, YoloX_T, YoloX_S, YoloX_M, YoloX_L, YoloX_X
+from super_gradients.training.utils.utils import HpmStruct
+
+
+class TestYOLOX(unittest.TestCase):
```
Comment (ofrimasad) [inline tests/unit_tests/yolox_unit_test.py:9]: there are many other tests i can think of. but let's keep them for next pr

File: src/super_gradients/training/losses/yolox_loss.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+            loss = loss.mean()
+        elif self.reduction == "sum":
+            loss = loss.sum()
+
+        return loss
+
+
+class YoloXDetectionLoss(_Loss):
+    """
+    Calculate YOLO V5 loss:
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:67]: fix typo v5

File: src/super_gradients/training/losses/yolox_loss.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+            predictions = model_output
+
+        return self.compute_loss(predictions, targets)
+
+    @staticmethod
+    def _make_grid(nx=20, ny=20):
+        yv, xv = torch.meshgrid([torch.arange(ny), torch.arange(nx)])
+        return torch.stack((xv, yv), 2).view((1, 1, ny, nx, 2)).float()
+
+    def compute_loss(self, predictions: List[torch.Tensor], targets: torch.Tensor) \
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:97]: please make this a private function

File: src/super_gradients/training/losses/yolox_loss.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+            if num_gt == 0:
+                cls_target = transformed_outputs.new_zeros((0, self.num_classes))
+                reg_target = transformed_outputs.new_zeros((0, 4))
+                l1_target = transformed_outputs.new_zeros((0, 4))
+                obj_target = transformed_outputs.new_zeros((total_num_anchors, 1))
+                fg_mask = transformed_outputs.new_zeros(total_num_anchors).bool()
+            else:
+                # GT boxes to image coordinates
+                gt_bboxes_per_image = labels_im[:, 2:6].clone()
+                # gt_bboxes_per_image = labels_im[:, 2:6] * self.im_size
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:156]: is this a comment or should be deleted?

File: src/super_gradients/training/losses/yolox_loss.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+
+class YoloXDetectionLoss(_Loss):
+    """
+    Calculate YOLO V5 loss:
+    L = L_objectivness + L_iou + L_classification + 1[no_aug_epoch]*L_l1
+    """
+
+    def __init__(self, strides, im_size, num_classes, use_l1=False, center_sampling_radius=2.5, iou_type='iou'):
+        super().__init__()
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:72]: please add pep8 docs

File: src/super_gradients/training/losses/yolox_loss.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+        if self.use_l1:
+            loss_l1 = self.l1_loss(raw_outputs.view(-1, 4)[fg_masks], l1_targets).sum() / num_fg
+        else:
+            loss_l1 = 0.0
+
+        reg_weight = 5.0
+        loss = reg_weight * loss_iou + loss_obj + loss_cls + loss_l1
+
+        return loss, torch.cat((loss_iou.unsqueeze(0), loss_obj.unsqueeze(0), loss_cls.unsqueeze(0),
+                                torch.tensor(loss_l1).unsqueeze(0).cuda(),
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:219]: do not use .cuda() 
use to(device) instead

File: src/super_gradients/training/losses/yolox_loss.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+            * L_objectivness is calculated for all cells.
+
+        L_classification:
+            for cells that have suitable ground truths in their grid locations add BCEs
+            to force a prediction of IoU with a GT in a multi-label way
+            Coef: 1.
+        L_iou:
+            for cells that have suitable ground truths in their grid locations
+            add (1 - IoU^2), IoU between a predicted box and each GT box, force maximum IoU
+            Coef: 1.
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:112]: looks like coef is actually 5.0 (line215)

File: src/super_gradients/training/losses/yolox_loss.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+        if mode == "cpu":
+            cls_preds_, obj_preds_ = cls_preds_.cpu(), obj_preds_.cpu()
+
+        with torch.cuda.amp.autocast(enabled=False):
+            cls_preds_ = cls_preds_.float().unsqueeze(0).repeat(num_gt, 1, 1).sigmoid_() * \
+                         obj_preds_.float().unsqueeze(0).repeat(num_gt, 1, 1).sigmoid_()
+            pair_wise_cls_loss = F.binary_cross_entropy(cls_preds_.sqrt_(), gt_cls_per_image, reduction="none").sum(-1)
+        del cls_preds_
+
+        cost = pair_wise_cls_loss + 3.0 * pair_wise_ious_loss + 100000.0 * (~is_in_boxes_and_center)
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:345]: please make 3.0 and 100000.0 parameters with a default value (set in the init)

File: src/super_gradients/training/losses/yolox_loss.py [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+
+        if mode == "cpu":
+            gt_matched_classes = gt_matched_classes.cuda()
+            fg_mask = fg_mask.cuda()
+            pred_ious_this_matching = pred_ious_this_matching.cuda()
+            matched_gt_inds = matched_gt_inds.cuda()
+
+        return gt_matched_classes, fg_mask, pred_ious_this_matching, matched_gt_inds, num_fg
+
+    def get_in_boxes_info(self, gt_bboxes_per_image, expanded_strides, x_shifts, y_shifts, total_num_anchors, num_gt):
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:361]: add pep 8 docs

File: src/super_gradients/training/losses/yolox_loss.py [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+        center_deltas = torch.stack([c_l, c_t, c_r, c_b], 2)
+        is_in_centers = center_deltas.min(dim=-1).values > 0.0
+        is_in_centers_all = is_in_centers.sum(dim=0) > 0
+
+        # in boxes OR in centers
+        is_in_boxes_anchor = is_in_boxes_all | is_in_centers_all
+
+        # in boxes AND in centers, preserving a shape [num_GTs x num_FGs]
+        is_in_boxes_and_center = (is_in_boxes[:, is_in_boxes_anchor] & is_in_centers[:, is_in_boxes_anchor])
+        return is_in_boxes_anchor, is_in_boxes_and_center
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:425]: I think can write this entire function with 4 lines using `torchvision.utils.draw_bounding_boxes` and `torchvision.utils.draw_keypoints'

**Review by ofrimasad — COMMENTED** [review]
1. please make all internal functions private (_ prefix)
2. add pep 8 docs where needed
3. you are assuming all tonsors are in GPU calling  .cuda(). that is risky. let's make them .to(device)

File: src/super_gradients/training/utils/detection_utils.py [hunk 12]
```diff
@@ -1164,3 +1165,21 @@ def num_anchors(self) -> int:
...
+    """
+    Collate function for Yolox training
+    """
+    def __call__(self, data):
+        batch = default_collate(data)
+        ims = batch[0]
+        targets = batch[1]
+        nlabel = (targets.sum(dim=2) > 0).sum(dim=1)  # number of objects
+        targets_merged = []
+        for i in range(targets.shape[0]):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/utils/detection_utils.py:1180]: For compatabality with our MAP implementation (also used by yolov5 which does it this way...so I rather just change it here then change a lot of parts that soon will be out of our code)

File: src/super_gradients/training/utils/detection_utils.py [hunk 13]
```diff
@@ -1164,3 +1165,21 @@ def num_anchors(self) -> int:
 
     def __repr__(self):
         return f"anchors_list: {self.__anchors_list} strides: {self.__strides}"
+
+
+class YoloXCollateFN:
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/utils/detection_utils.py:1170]: sure

File: tests/unit_tests/yolox_unit_test.py (hunk 1 above)
Comment (shaydeci) [reply tests/unit_tests/yolox_unit_test.py:43]: done

File: src/super_gradients/training/losses/yolox_loss.py (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py:156]: deleted (used to have unnormalised labels earlier on...)

File: src/super_gradients/training/losses/yolox_loss.py (hunk 7 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py:219]: done

File: src/super_gradients/training/losses/yolox_loss.py (hunk 8 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py]: fixed

Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py]: fixed

File: src/super_gradients/training/losses/yolox_loss.py (hunk 9 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py]: done

File: src/super_gradients/training/losses/yolox_loss.py (hunk 11 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py:500]: Don't see how drawing boxes / key points on top of an image would help here....Maybe now that I added docs the purpose of this method is more clear.

**Comment by yurkovak** [comment]
TODO: integrate changes from #196

File: src/super_gradients/training/models/detection_models/yolov5_base.py [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+    'nms_iou': 0.45,  # When add_nms is True IoU threshold for NMS algorithm
+    # (with smaller value more boxed will be considered "the same" and removed)
+    'yolo_type': 'yoloV5',  # Type of yolo to build: 'yoloV5' and 'yoloX' are supported
+    'yolo_version': 'v6.0',  # Release version of Ultralytics yoloV5 to build a model from: v6.0 and v3.0 are supported
+                             # (has an impact only if yolo_type is yoloV5)
+    'stem_type': None,  # 'focus' and '6x6' are supported, by default is defined by yolo_type and yolo_version
+}
+
+
+class YoloV5PostPredictionCallback(DetectionPostPredictionCallback):
```
Comment (yurkovak) [inline src/super_gradients/training/models/detection_models/yolov5_base.py:47]: Check out most of the changed files in #192. There all references to yolov5 are renamed to yolov5_base all over the code. The most sensitive part is the import of `YoloV5PostPredictionCallback` which we have a lot of, including in notebooks. Now it should be imported from yolov5_base

Comment (shaydeci) [reply src/super_gradients/training/models/detection_models/yolov5_base.py:47]: Taken care of in https://github.com/Deci-AI/super-gradients/pull/207 . Shouldn't be relevant once we update the branch.

**Comment by shaydeci** [comment]
> TODO: integrate changes from #196

Integrated in https://github.com/Deci-AI/super-gradients/pull/210/commits/e8cc8de1c0d37b5890718d9e454ae763052ff5a3

File: src/super_gradients/training/models/detection_models/csp_darknet53.py [hunk 15]
```diff
@@ -215,18 +233,23 @@ def __init__(self, arch_params: HpmStruct):
...
             raise NotImplementedError(f'One of {yolo_type} yolo type or {yolo_version} yolo version is not supported')
 
-        self._modules_list.append(Conv(width_mult(64), width_mult(128), 3, 2, activation_type))                      # 1
-        self._modules_list.append(block(width_mult(128), width_mult(128), struct[0], activation_type))               # 2
-        self._modules_list.append(Conv(width_mult(128), width_mult(256), 3, 2, activation_type))                     # 3
-        self._modules_list.append(block(width_mult(256), width_mult(256), struct[1], activation_type))               # 4
-        self._modules_list.append(Conv(width_mult(256), width_mult(512), 3, 2, activation_type))                     # 5
-        self._modules_list.append(block(width_mult(512), width_mult(512), struct[2], activation_type))               # 6
-        self._modules_list.append(Conv(width_mult(512), width_mult(1024), 3, 2, activation_type))                    # 7
+        self._modules_list.append(ConvBlock(width_mult(64), width_mult(128), 3, 2, activation_type))                 # 1
```
Comment (ofrimasad) [inline src/super_gradients/training/models/detection_models/csp_darknet53.py:236]: this part of the code really looks like it should have been a for loop (lines 236 to 245)

File: src/super_gradients/training/losses/yolox_loss.py [hunk 16]
```diff
@@ -0,0 +1,534 @@
+"""
+Based on https://github.com/Megvii-BaseDetection/YOLOX
+"""
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py:4]: Added.

File: src/super_gradients/training/losses/yolox_loss.py [hunk 17]
```diff
@@ -0,0 +1,534 @@
...
+            * 'giou' according to "Generalized Intersection over Union: A Metric and A Loss for Bounding Box Regression"
+                (1 - giou), where giou = iou - (cover_box - union_box)/cover_box
+    """
+
+    def __init__(self, reduction: str = "none", loss_type: str = "iou"):
+        super(IOUloss, self).__init__()
+        if loss_type not in ["iou", "giou"]:
+            raise ValueError("Illegal loss_type value: " + loss_type)
+        if reduction not in ["mean", "sum", "none"]:
+            raise ValueError("Illegal reduction value: " + reduction)
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py]: Added in  [license and validity method added to yolox loss](https://github.com/Deci-AI/super-gradients/pull/210/commits/13b40aea02f859cfac67d72ea4967b5ec761d0d0)

File: src/super_gradients/training/models/detection_models/csp_darknet53.py (hunk 15 above)
Comment (shaydeci) [reply src/super_gradients/training/models/detection_models/csp_darknet53.py:236]: Added in [license and validity method added to yolox loss](https://github.com/Deci-AI/super-gradients/pull/210/commits/13b40aea02f859cfac67d72ea4967b5ec761d0d0)

File: src/super_gradients/training/losses/yolox_loss.py [hunk 18]
```diff
@@ -0,0 +1,534 @@
...
+            * 'iou' for
+                (1 - iou^2)
+            * 'giou' according to "Generalized Intersection over Union: A Metric and A Loss for Bounding Box Regression"
+                (1 - giou), where giou = iou - (cover_box - union_box)/cover_box
+    """
+
+    def __init__(self, reduction: str = "none", loss_type: str = "iou"):
+        super(IOUloss, self).__init__()
+        if loss_type not in ["iou", "giou"]:
+            raise ValueError("Illegal loss_type value: " + loss_type)
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py]: Added in  [license and validity method added to yolox loss](https://github.com/Deci-AI/super-gradients/pull/210/commits/13b40aea02f859cfac67d72ea4967b5ec761d0d0)

File: src/super_gradients/training/models/detection_models/csp_darknet53.py (hunk 15 above)
Comment (shaydeci) [reply src/super_gradients/training/models/detection_models/csp_darknet53.py]: done.

**Review by ofrimasad — APPROVED** [review]
LGTM
